integration/update_compat_test.go

```diff
@@ -71,96 +75,177 @@ func testUpdateCompat(t *testing.T, testCases map[string]updateCompatTestCase) {
 				Providers:                providers,
 				AddNonExistentCollection: true,
 			})
-			ctx, targetCollections, compatCollections := s.Ctx, s.TargetCollections, s.CompatCollections
 
-			update, replace := tc.update, tc.replace
-			if update != nil {
-				require.Nil(t, replace, "`replace` must be nil if `update` is set")
-			} else {
-				require.NotNil(t, replace, "`replace` must be set if `update` is nil")
-			}
+			nonEmptyResults := testUpdateCollections(t, s, updateCollectionsParams{
+				update:  tc.update,
+				replace: tc.replace,
+				filter:  tc.filter,
+			})
 
-			var nonEmptyResults bool
-			for i := range targetCollections {
-				targetCollection := targetCollections[i]
-				compatCollection := compatCollections[i]
-				t.Run(targetCollection.Name(), func(t *testing.T) {
-					t.Helper()
+			assert.True(t, nonEmptyResults, "expected non-empty results (some documents should be modified)")
+		})
+	}
+}
 
-					allDocs := FindAll(t, ctx, targetCollection)
+// testUpdateCollections applies updateOne operation to collections
+// and returns true if any collection was modified.
+func testUpdateCollections(t *testing.T, s *setup.SetupCompatResult, p updateCollectionsParams) bool {
```

Comment:
testUpdateCollections captures common part of `testUpdateCompat` and `testUpdateCompatUnchanged`